Assets/LeapMotionModules/UI/Interaction/InteractionHand.cs

```diff
@@ -0,0 +1,780 @@
+        }
+
+        if (inCurFrame && !inLastFrame) {
+          hoverable.HoverBegin(hoverResults.checkedHand);
+          _hoveredLastFrame.Add(hoverable);
+        }
+        if (inCurFrame && inLastFrame) {
+          hoverable.HoverStay(hoverResults.checkedHand);
+        }
+        if (!inCurFrame && inLastFrame) {
+          hoverable.HoverEnd(hoverResults.checkedHand);
+          _hoveredLastFrame.Remove(hoverable);
+        }
+      }
+
+      foreach (var hoverable in _hoveredLastFrame) {
+        if (!trackedBehaviours.Contains(hoverable)) {
+          hoverable.HoverEnd(hoverResults.checkedHand);
+          _removalCache.Add(hoverable);
+        }
+      }
+      foreach (var hoverable in _removalCache) {
+        _hoveredLastFrame.Remove(hoverable);
+      }
+      _removalCache.Clear();
+    }
+
+    private void ProcessPrimaryHoverCheckResults(HoverCheckResults hoverResults) {
+      if (hoverResults.primaryHovered == _primaryHoveredLastFrame) {
+        if (hoverResults.primaryHovered != null) hoverResults.primaryHovered.PrimaryHoverStay(hoverResults.checkedHand);
+      }
+      else {
+        if (_primaryHoveredLastFrame != null) {
+          _primaryHoveredLastFrame.PrimaryHoverEnd(hoverResults.checkedHand);
+        }
+        _primaryHoveredLastFrame = hoverResults.primaryHovered;
+        if (_primaryHoveredLastFrame != null) _primaryHoveredLastFrame.PrimaryHoverBegin(hoverResults.checkedHand);
+      }
+    }
+
+    private const float MAX_PRIMARY_HOVER_DISTANCE = 0.1F;
+    private Vector3 _intentionPosition;
+    private InteractionBehaviourBase _primaryHoveredLastFrame = null;
+    private float[] _intentionFingerWeights = new float[NUM_FINGERS] { 0F, 1F, 0.6F, 0.4F, 0.1F };
+
+    // Note. This is used as the hover sphere-check origin, but distance data is not calculated from it.
+    // If it's worthwhile, it should be used to provide distance in hover callbacks... otherwise it should just be cut.
+    private void CalculateIntentionPosition() {
+      if (_hand == null) return;
+
+      // Weighted average of medial finger bone positions
+      float usageSum = 0F;
+      Leap.Vector averagePosition = Leap.Vector.Zero;
+      //Leap.Vector averageDirection = Leap.Vector.Zero;
+      for (int i = 1; i < 5; i++) {
+        Leap.Vector distalFingerBoneTip = _hand.Fingers[i].bones[3].NextJoint;
+        Leap.Vector medialFingerBoneDirection = _hand.Fingers[i].bones[2].Direction;
+        Leap.Vector distalDirection = _hand.Basis.zBasis;
+        var fingerUsage = (((medialFingerBoneDirection.x * distalDirection.x)
+                         + (medialFingerBoneDirection.y * distalDirection.y)
+                         + (medialFingerBoneDirection.z * distalDirection.z))
+                         * _intentionFingerWeights[i])
+                         .Map(-0.3F, 0.9F, 0, 1);
+        usageSum += fingerUsage;
+        averagePosition = new Leap.Vector(averagePosition.x + distalFingerBoneTip.x * fingerUsage,
```

Comment:
Any reason you are doing the component math manually instead of just doing
`averagePosition + distalFingerBoneTip * fingerUsage`
?